cirq/protocols/approx_equality.py

```diff
@@ -0,0 +1,172 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+from typing import Any
+
+from typing_extensions import Protocol
+
+
+class SupportsApproxEquality(Protocol):
+    """Object which can be approximately compared."""
+
+    def _approx_eq_(self, other: Any, rel_tol: float, abs_tol: float) -> bool:
+        """Approximate comparator.
+
+        Types implementing this protocol define their own logic for approximate
+        comparison with other types.
+
+        Args:
+        other: Target object for approximate comparison.
+        rel_tol: The relative tolerance. See math.isclose() documentation for
+            details.
+        abs_tol: The minimum absolute tolerance. See math.isclose()
+            documentation for details.
+
+        Returns:
+            True if objects are approximately equal, False otherwise. Returns
+            NotImplemented when approximate equality is not implemented for
+            given types.
+        """
+        pass
+
+
+def approx_eq(
+        val: Any,
+        other: Any,
+        rel_tol: float = 1e-09,
+        abs_tol: float = 0.0) -> bool:
+    """Approximately compares two objects.
+
+    If `val` implements SupportsApproxEquality protocol then it is invoked and
+    takes precedence over all other checks:
+     - For primitive numeric types `int` and `float` approximate equality is
+       delegated to math.isclose().
+     - For complex primitive type the real and imaginary parts are treated
+       independently and compared using math.isclose().
+     - For `val` and `other` both iterable, consecutive elements are compared
+       recursively. Types of `val` and `other` does not necessarily needs to
+       match each other. They just need to be iterable and have the same
+       structure.
+
+    Args:
+        val: Source object for approximate comparison.
+        other: Target object for approximate comparison.
+        rel_tol: The relative tolerance. See math.isclose() documentation for
+            details.
+        abs_tol: The minimum absolute tolerance. See math.isclose()
+            documentation for details.
+
+    Returns:
+        True if objects are approximately equal, False otherwise. Returns
+        NotImplemented when approximate equality is not implemented for given
+        types.
+    """
+
+    # Check if object defines approximate equality via _approx_eq_. This takes
+    # precedence over all other overloads.
+    approx_eq_getter = getattr(val, '_approx_eq_', None)
+    if approx_eq_getter is not None:
+        return approx_eq_getter(other, rel_tol, abs_tol)
+
```

Comment:
Because approximate equality is symmetric, we should also check `getattr(other, '_approx_eq_', None)` before continuing to the default cases. For example, `cirq.approx_eq(5.0, fraction(5, 1))` should work.

Reply:
Done.